### DISCUSSION TITLE
Encountering mode

### DISCUSSION THREAD

**Discussion by KnorpelSenf** [telegraf/telegraf#1218] [Ideas]
I have a suggestion. It is based on https://github.com/telegraf/telegraf/pull/1195#issuecomment-735421884 and could work as follows:

```ts
// some middleware, just like before
const mw = ctx => {}

// exactly one update type
bot.sees('message' mw) // all message updates
bot.sees('channelPost', mw) // all channel_post updates  (1)
bot.sees(['poll', 'pollAnswer'], mw) // all poll-related updates

// one update type of a subtype
bot.sees('message:text', mw) // all message updates with text
bot.sees('callbackQuery:data', mw) // all callback_query updates with data
bot.sees('editedMessage:photo', mw) // all edited_message updates of messages containing a photo

// MessageEntities and Users can be filtered even further  (2)
bot.sees('message:entities:url', mw) // all updates with messages that contain a URL
bot.sees('editedChannelPost:captionEntities:italic', mw) // all updates about edits to channel posts that contain italic text in the caption
bot.sees('leftChatMember:is_bot', mw) // all updates about a bot that left the chat

// leaving out the part before the colon makes reasonable assumptions
bot.sees(':forward', mw) // all updates that are forwarded messages (effectively equal to `message:forward` and `channelPost:forward`)
bot.sees([':text', ':caption'], mw) // all updates that contain text or a caption, except edits
bot.sees('::cashtag', mw) // all updates that have a cashtag in either text or caption of either message or channelPost, except edits

// useful shortcuts 
bot.sees('edit', mw) // all updates that are edits
bot.sees('edit:text', mw) // all updates that are edits to text messages
bot.sees(':new_chat_members:me', mw) // all updates about the bot itself joining a chat  (3)
```

All methods are fail-fast. They throw immediately when called with wrong syntax, rather than performing the matching in an unexpected way.

There are a few things we need to discuss.

#### About the Annotations in the Example

Regarding the annotations `(1)`, `(2)`, and `(3)` in the code, I have the following things to say.
1. I suggest to use camelCase rather than snake_case, e.g. `channelPost` and not `channel_post`. It does introduce odd renaming, but it is consistent with JS/TS naming conventions as well as the previous names, i.e. `ctx.channelPost`. We should be clear about this in the docs.
2. This could be done in the 4.1 release and skipped for now.
3. Do we even want this? Could also be postponed until 4.1.

#### Naming
This syntax is not compatible with `bot.on`, e.g. `bot.on('text')` would need to be expressed by `bot.sees(':text')`. I think it is the right thing to introduce a new method rather than changing the old one, as that would most likely break 98 % of bots ever written.

Is there a better name than `sees`? I thought about `for`, `with`, `receives`, and `gets`, but I don't like any of them.

#### Implementation Cost
Even though this syntax looks advanced, it should in fact be relatively simple to implement. I'm willing to look into this (no time frame promised) if we agree that it is a good idea.

I have an idea how to keep up performance, even though updates like `'::url'` do require some searching.

_Originally posted by @KnorpelSenf in https://github.com/telegraf/telegraf/issues/1202#issuecomment-736344271_

**Comment by KnorpelSenf** [comment]
The only problem I see so far with `bot.sees` is that you cannot specify multiple filters, e.g. it is impossible to get all forwarded text messages. Forwards are caught by `message:forward` and text messages by `message:text`, but you cannot combine them.

One solution would be to allow multiple entries between the colons, akin to `message:text,forward` or `message:text&forward`, but this would look very odd with having a third option that filters the second further down. What would `message:forward,entities,text:url` mean? Even it technically does not clash because the Telegram API is coincidentally structured in such a way that we could always figure out which of the second options the `url` would refer to, it look so confusing that I'm not willing to implement it.

Another way to tackle this is by allowing nested arrays. As an example, `['message:text', 'channelPost:text']` would give all texts in either messages or channel posts, and `[['message:text', 'message:forward'], ['channelPost:text']]` would additionally include all those messages that are forwards. You could basically use a set notation for a logical formula in DNF, which would be consistent with the existing methods.

I think the most promising solution would be a combination of the two.
```ts
bot.sees('message:text&message:forward', mw)
bot.sees('::url&:text', mw)
```
Any number of filters can be appended, separated by `&`. If we allow whitespace around the separator, this would make it even more readable.

```ts
bot.sees('message:text & message:forward', mw)
bot.sees('::url & :text', mw)
```

Supporting any number of filters certainly makes it more complicated to implement than the original idea. However, it is also by far the most intuitive notation for such advanced yet concise filtering.

I'm happy to hear what other people think about this. This feature will not be ready for the 4.0 release in #1202, but it could ship in the minor version after that.

**Comment by wojpawlik** [comment]
Keep in mind that TypeScript types need to be able to understand the string as well to provide high-quality types. How would they understand anything more complex that `bot.selector('edited_message.video_note')`?

**Reply by KnorpelSenf** [reply]
I'm convinced that this is possible in theory with string literal types and recursive conditional types, so we can somehow recursively intersect stuff to derive the correct type for `guard`. However, I do not have an implementation yet, so yes, it may turn out to be so impractical/slow/much work that we have to drop it. (Or I could be wrong with my assumption, too.)

**Comment by wojpawlik** [comment]
I have a good initial implementation in mind, I can draft PR after #1257. We can settle on the name in the meantime. I suggest `Composer.select`, because the examples with entity matching remind me of CSS selectors.
